### DISCUSSION TITLE
Only CLIENT_RANDOM SSLKEYLOGFILE for curl command

### DISCUSSION THREAD

**Discussion by NanFeng009** [curl/curl#9617] [Q&A] [answered]
Use Curl command line with environment SSLKEYLOGFILE  and only  found
CLIENT_RANDOM xxxxxxxxxxxxxx xxxxxxxxxxxxxxxx saved. why there is no other information like below. Anyone has idea?
SERVER_HANDSHAKE_TRAFFIC_SECRET
EXPORTER_SECRET
SERVER_TRAFFIC_SECRET_0
CLIENT_HANDSHAKE_TRAFFIC_SECRET
CLIENT_TRAFFIC_SECRET_0

**Comment by NanFeng009** [comment]
To decrypt the traffic, i need all these keys. only CLIENT_RANDOM, only very little information can be decrypted.
I tried using firefox, and can see all these information in SSLKEYLOGFILE. is there any special setting need for CURL?

**Comment by bagder** [comment]
You failed to tell us what curl version you are using and what TLS backend. Your `curl -V` output for example.

**Reply by NanFeng009** [reply]
curl --version
curl 7.81.0 (x86_64-pc-linux-gnu) libcurl/7.81.0 OpenSSL/3.0.2 zlib/1.2.11 brotli/1.0.9 zstd/1.4.8 libidn2/2.3.2 libpsl/0.21.0 (+libidn2/2.3.2) libssh/0.9.6/openssl/zlib nghttp2/1.43.0 librtmp/2.3 OpenLDAP/2.5.13
Release-Date: 2022-01-05
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets zstd

**Comment by jay** [comment, marked as answer]
Those other symbols are used by TLS 1.3. If the connection does not use TLS 1.3 (for example you are using an older version of OpenSSL that does support TLS 1.3) then probably only CLIENT_RANDOM will be output, but that should be all you need.

**Reply by NanFeng009** [reply]
I used Ubuntu 2204 and Openssl 3.0 should be default. and libssl is  strings /lib/x86_64-linux-gnu/libssl.so.3 |grep -i "^OpenSSL_[0-9]"
OPENSSL_3.0.0
OPENSSL_3.0.3
But i saw from terminal TLS 1.2 was really negotiated, it seems align with your answer.
Curl use libssl and libssl will record these information(EXPORTER_SECRET) in and after TLS 1.3. So this issue is more related with libssl, am i right?

**Reply by jay** [reply]
Yes. If libcurl was built to use OpenSSL 1.1.1 or later then it will call SSL_CTX_set_keylog_callback to set a callback which receives the secrets, and that callback should write everything that it's given. As I've explained secrets such as EXPORTER_SECRET are only for TLS 1.3.

https://github.com/curl/curl/blob/93d092867f0f2c78571983040ef75e078ee1a4c4/lib/vtls/openssl.c#L286-L293

https://github.com/curl/curl/blob/93d092867f0f2c78571983040ef75e078ee1a4c4/lib/vtls/keylog.c#L87-L114

**Comment by NanFeng009** [comment]
Really appreciate your answer👍.  This has solved my long time puzzle.
